### PR TITLE
Add support for aws auth backend for auth_type=iam

### PR DESCRIPTION
**Pull request by shayangz**
Submitting a PR for adding aws iam auth method (released with Vault 0.7.1)

### CODE REVIEW

File: vault.gemspec [hunk 1]
```diff
@@ -19,6 +19,8 @@ Gem::Specification.new do |spec|
   spec.executables   = spec.files.grep(%r{^exe/}) { |f| File.basename(f) }
   spec.require_paths = ["lib"]
 
+  spec.add_runtime_dependency "aws-sdk"
```
Comment (sethvargo) [inline vault.gemspec]: We keep the client libraries to a zero runtime dependency requirement to make them fast, flexible, and unlikely to conflict with other libraries. Given how heavy the aws-sdk is, I am very opposed to including it as a dependency.

Comment (shayangz) [inline vault.gemspec]: Makes sense. 

Do you have a recommendation for how to generate the aws signatures required for the new auth endpoint short of re-implementing a significant amount of logic?

Comment (sethvargo) [inline vault.gemspec]: It _should_ just be hmacing a bunch of things.

```ruby
def aws_signature(key, timestamp, region, service)
  k_date    = OpenSSL::HMAC.digest('sha256', "AWS4" + key, timestamp)
  k_region  = OpenSSL::HMAC.digest('sha256', k_date, region)
  k_service = OpenSSL::HMAC.digest('sha256', k_region, service)
  k_signing = OpenSSL::HMAC.digest('sha256', k_service, "aws4_request")
  kSigning
end
```

This looks to be a more complete example: https://gist.github.com/blelump/9e6880a87fefa6397032

**Comment by sethvargo** [comment]
Hey @shayangz 

Did you have a chance to revisit this?

**Comment by shayangz** [comment]
Unfortunately haven't had a chance. It's a bit more involved than just doing  a bunch of HMACs if we want to properly support different AWS auth methods (instance profiles, etc) which I think is necessary for this type of experience.

**Comment by jhmartin** [comment]
Is the mispelling of `iam_auth_header_vaule` intentional?

**Comment by shayangz** [comment]
@jhmartin no, that's a bug 🤦‍♂️.

**Comment by tristanmorgan** [comment]
Hi @shayangz , I created a gist with an attempt to get a vault token via IAM credentials. might give you some hints on how it could work.
https://gist.github.com/tristanmorgan/8fe883f89db4537ce35e0a23f6a537b6

**Comment by coen-hyde** [comment]
AWS has broken up their monolithic aws-sdk gem into a hundred pieces. One of the broken out gems is `aws-sigv4`. Which should be all you need right?

**Comment by evanphx** [comment]
I'd be willing to take a look at the version of this PR that only used the `aws-sigv4` gem.

**Comment by evanphx** [comment]
This happened in #161 instead!

**Pull request closed without merging**